metadata/client.go

```diff
@@ -0,0 +1,112 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metadata
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+)
+
+import (
+	"github.com/dubbogo/gost/log/logger"
+)
+
+import (
+	"dubbo.apache.org/dubbo-go/v3/common"
+	"dubbo.apache.org/dubbo-go/v3/common/constant"
+	"dubbo.apache.org/dubbo-go/v3/common/extension"
+	"dubbo.apache.org/dubbo-go/v3/metadata/info"
+	reportInstance "dubbo.apache.org/dubbo-go/v3/metadata/report/instance"
+	"dubbo.apache.org/dubbo-go/v3/registry"
+)
+
+const metadataProxyDefaultTimeout = 5000
+
+// GetMetadataFromMetadataReport test depends on dubbo protocol, if dubbo not dependent on config package, can move to metadata dir
+func GetMetadataFromMetadataReport(revision string, instance registry.ServiceInstance) (*info.MetadataInfo, error) {
+	report := reportInstance.GetMetadataReport()
+	return report.GetAppMetadata(instance.GetServiceName(), revision)
+}
+
+func GetMetadataFromRpc(revision string, instance registry.ServiceInstance) (*info.MetadataInfo, error) {
+	service, destroy := createRpcClient(instance)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(metadataProxyDefaultTimeout))
+	defer cancel()
+	defer destroy()
+	return service.GetMetadataInfo(ctx, revision)
+}
+
+type remoteMetadataService struct {
+	//GetExportedURLs       func(context context.Context, serviceInterface string, group string, version string, protocol string) ([]*common.URL, error) `dubbo:"getExportedURLs"`
+	GetMetadataInfo func(context context.Context, revision string) (*info.MetadataInfo, error) `dubbo:"getMetadataInfo"`
+	//GetMetadataServiceURL func(context context.Context) (*common.URL, error)
+	//GetSubscribedURLs     func(context context.Context) ([]*common.URL, error)
+	//Version               func(context context.Context) (string, error)
+}
+
+func createRpcClient(instance registry.ServiceInstance) (*remoteMetadataService, func()) {
+	params := getMetadataServiceUrlParams(instance.GetMetadata()[constant.MetadataServiceURLParamsPropertyName])
+	url := buildMetadataServiceURL(instance.GetServiceName(), instance.GetHost(), params)
+	return createRpcClientByUrl(url)
+}
+
+func createRpcClientByUrl(url *common.URL) (*remoteMetadataService, func()) {
+	rpcService := &remoteMetadataService{}
+	invoker := extension.GetProtocol(constant.Dubbo).Refer(url)
+	proxy := extension.GetProxyFactory("").GetProxy(invoker, url)
```

Comment:
```extension.GetProxyFactory("")``` works relying on that GetProxyFactory would treat empty string as "default". Maybe using "default" directly is better.